### PR TITLE
Limit length of accepted email addresses

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: synapse/util/threepids.py [hunk 1]
```diff
@@ -18,6 +18,16 @@
 logger = logging.getLogger(__name__)
 
 
+# it's unclear what the maximum length of an email address is. RFC3696 (as corrected
+# by errata) says:
+#    the upper limit on address lengths should normally be considered to be 254.
+#
+# In practice, mail servers appear to be more tolerant and allow 400 characters
+# or so. Let's allow 500, which should be plenty for everyone.
```
Comment (clokep) [inline synapse/util/threepids.py:26]: Is this for the full email address (including the domain name)?

Comment (erikjohnston) [reply synapse/util/threepids.py:26]: FWIW I stole this from sydent https://github.com/matrix-org/sydent/blob/c2ab10897500dbba1c0d4be33f42985d81bbad7d/sydent/util/stringutils.py#L29-L36

I don't know if it includes a domain name, but also :shrug:

File: synapse/util/threepids.py [hunk 2]
```diff
@@ -70,3 +80,15 @@ def canonicalise_email(address: str) -> str:
...
+def validate_email(address: str):
+    """Does some basic validation on an email address.
+
+    Raises a ValueError if the email is invalid.
+    """
+    # First we try canonicalising in case that fails
+    address = canonicalise_email(address)
+
+    if len(address) > MAX_EMAIL_ADDRESS_LENGTH:
+        raise ValueError("Unable to parse email address")
```
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/util/threepids.py:94]: Ah, I see that `canonicalise_email` ensures there's exactly a single `@` in it.

**Review by clokep — COMMENTED** [review]
Looks pretty plausible!

File: synapse/util/threepids.py [hunk 3]
```diff
@@ -70,3 +80,15 @@ def canonicalise_email(address: str) -> str:
...
     return parts[0].casefold() + "@" + parts[1].lower()
+
+
+def validate_email(address: str):
+    """Does some basic validation on an email address.
+
+    Raises a ValueError if the email is invalid.
+    """
+    # First we try canonicalising in case that fails
+    address = canonicalise_email(address)
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/util/threepids.py:91]: Yeah good point.

**Review by clokep — APPROVED** [review]
LGTM.

**Pull request merged**